### PR TITLE
implement ethereum-compatible websockets server

### LINKED ISSUE #355: implement ethereum JSON-RPC compatible websockets server
__Proposal:__ [Description of the feature]

currently, ethermint has tendermint's websockets server enabled, which can be used as follows: https://docs.tendermint.com/master/rpc/ it can be connected to at `ws://localhost:26657/websocket` when running a node.

however, this doesn't conform to ethereum's pubsub JSON-RPC API, see https://geth.ethereum.org/docs/rpc/pubsub

we should either extend the existing rest server (not sure of feasibility of this) or create another server that supports ws that implements go ethereum's pubsub API.

__Current behavior:__ [What currently happens]

websockets API is tendermint's, doesn't conform to ethereum

__Desired behavior:__ [What you would like to happen]

able to connect to a websockets server and get ethermint events using ethereum's API

__Use case:__ [Why is this important (helps with prioritizing requests)]

Requests may be closed if we're not actively planning to work on them.

### PR DESCRIPTION
**Pull request by noot**
Closes: #355

## Description

- [x] start a websockets server when the rest-server starts 
- [x] begin ethereum pubsub api (eth_subscribe, eth_unsubscribe)
- [x] allow subscriptions for `newHeads`
- [x] allow subscriptions for `logs`
- [x] allow subscriptions for `newPendingTransactions`
- [x] stub subscriptions for `syncing`

test instructions:
```
make install
./init.sh
``` 

in another terminal: (can optionally specify  `--wsport [port]` default is 8546)
```
emintcli rest-server --laddr "tcp://localhost:8545" --unlock-key mykey --chain-id 8 --trace
```

in another terminal, install wscat or use whichever ws client you like:
```
npm i -g wscat
wscat -c "ws://localhost:8546"
```

the usual eth api is supported:

```
> {"jsonrpc":"2.0","method":"eth_blockNumber","params":[],"id":1}
< {"id":1,"jsonrpc":"2.0","result":"0x2"}
```
for block subscriptions:
```
> {"id": 1, "method": "eth_subscribe", "params": ["newHeads", {}]}
< {"jsonrpc":"2.0","result":"0x44e010cb2c3161e9c02207ff172166ef","id":1}
```
you should see new block headers every couple seconds

for logs:
```
> {"id": 1, "method": "eth_subscribe", "params": ["logs"]}
< {"jsonrpc":"2.0","result":"0x4f2f0f10e345e3f5ddaa8da3d8eafc6f","id":1}
< {"jsonrpc":"2.0","method":"eth_subscription","params":{"subscription":"0x4f2f0f10e345e3f5ddaa8da3d8eafc6f","result":[{"address":"0xf025c23e267a7cba265e8baf627dd2cb1d010a44","topics":["0x775a94827b8fd9b519d36cd827093c664f93347070a554f65e4a6f56cd738898","0x0000000000000000000000000000000000000000000000000000000000000011"],"data":"0x","blockNumber":"0x1a","transactionHash":"0xcf07c0ff164121b2984209be151fee6fa2734fd11aca531240249e4dc80e5ecc","transactionIndex":"0x0","blockHash":"0x0000000000000000000000000000000000000000000000000000000000000000","logIndex":"0x0","removed":false}]}}
```
in another terminal send a tx that emits some logs: `MODE=rpc go test ./tests/ -run TestEth_GetTransactionLogs`
 you should see the logs in the ws console

can also handle logs subscription with criteria:
```
>> {"id": 1, "method": "eth_subscribe", "params": ["logs", {"address": "0x8320fe7702b96808f7bbc0d4a888ed1468216cfd", "topics": ["0xd78a0cb8bb633d06981248b816e7bd33c2a35a6089241d099fa519e361cab902"]}]}
<< {"jsonrpc":"2.0","id":2,"result":"0x4a8a4c0517381924f9838102c5a4dcb7"}
```

for pending txs:
```
> {"id": 1, "method": "eth_subscribe", "params": ["newPendingTransactions"]}
< {"jsonrpc":"2.0","result":"0x7127b43625439e8e92fd9f653fd490eb","id":1}
```
re-run step to send a tx, you should see the transaction in the ws console
```
< {"jsonrpc":"2.0","method":"eth_subscription","params":{"subscription":"0x7127b43625439e8e92fd9f653fd490eb","result":"0x23d4907a91f121f9418898ede340711978e714b4fe17ed7ae55b978c0c889284"}}
```
______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: rpc/websockets.go [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+	go func() {
+		err := http.ListenAndServe(fmt.Sprintf(":%s", s.wsAddr), ws)
+		if err != nil {
+			s.logger.Error("http error:", err)
+		}
+	}()
+}
+
+func (s *websocketsServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var upgrader = websocket.Upgrader{
```
Comment (fedekunze) [inline rpc/websockets.go:89]: q: should we fill in the other fields? check the ethereum implementation: https://github.com/ethereum/go-ethereum/blob/master/rpc/websocket.go#L49

File: rpc/websockets.go [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+		CheckOrigin: func(r *http.Request) bool {
```
Comment (fedekunze) [inline rpc/websockets.go:90]: we should consider updating it (maybe for a follow-up PR): https://github.com/ethereum/go-ethereum/blob/master/rpc/websocket.go#L69

File: rpc/websockets.go [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+		logger:  log.NewTMLogger(log.NewSyncWriter(os.Stdout)).With("module", "websocket-server"),
+	}
+}
+
+func (s *websocketsServer) start() {
+	ws := mux.NewRouter()
+	ws.Handle("/", s)
+
+	go func() {
+		err := http.ListenAndServe(fmt.Sprintf(":%s", s.wsAddr), ws)
```
Comment (fedekunze) [inline rpc/websockets.go:81]: Does it only support localhost?

File: rpc/config.go [hunk 4]
```diff
@@ -25,28 +25,15 @@ import (
 
 const (
 	flagUnlockKey = "unlock-key"
+	flagWebsocket = "websocket-port"
```
Comment (fedekunze) [inline rpc/config.go]: maybe we should use `wsport` since that's the one used on Geth

**Review by fedekunze — APPROVED** [review]

File: rpc/websockets.go (hunk 1 above)
Comment (noot) [reply rpc/websockets.go:89]: hmm, I'll investigate and do a follow up PR

File: rpc/websockets.go (hunk 3 above)
Comment (noot) [reply rpc/websockets.go:81]: the godocs say it can be an address, but in their examples they only have the port. I think it could support other addresses

**Pull request merged**